### PR TITLE
Add an example using codegen

### PR DESCRIPTION
**Pull request by mziolekda**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -39,6 +39,7 @@ To set a project up:
 
   * examples.pingpong.grpc.PingPongGrpcMain
   * examples.pingpong.reactive.PingPongReactiveMain
+  * examples.pingpong.codegen.PingPongCodegenMain
 
   depending on which example you wish to run. The script will take care of stopping an already running sandbox & start a fresh one every call.
```
Comment (atriantafyllos-da) [inline README.rst]:
```suggestion
  depending on which example you wish to run. The script will take care of stopping an already running sandbox & start a fresh one on every call.
```

File: README.rst [hunk 2]
```diff
@@ -64,22 +65,22 @@ The main function:
...
+======================
 
 The core of the application is the method `PingPongProcessor.runIndefinitely() <src/main/java/examples/pingpong/grpc/PingPongProcessor.java#L61-L91>`_.
 
 This method retrieves a gRPC streaming endpoint using the ``GetTransactionsRequest`` request, and then creates a `RxJava <The Underlying Library: RxJava_>`_ ``StreamObserver``, providing implementations of the ``onNext``, ``onError`` and ``onComplete`` observer methods. ``RxJava`` arranges that these methods receive stream events asynchronously.
 
-The method `onNext <src/main/java/examples/pingpong/grpc/PingPongProcessor.java#L74-L76>`_ is the main driver, extracting the transaction list from each ``GetTransactionResponse``, and passing in to  ``processTransaction()`` for processing. This method, and the method ``processTransaction()`` implents the application logic.
+The method `onNext <src/main/java/examples/pingpong/grpc/PingPongProcessor.java#L74-L76>`_ is the main driver, extracting the transaction list from each ``GetTransactionResponse``, and passing in to  ``processTransaction()`` for processing. This method, and the method ``processTransaction()`` implements the application logic.
```
Comment (atriantafyllos-da) [inline README.rst]:
```suggestion
The method `onNext <src/main/java/examples/pingpong/grpc/PingPongProcessor.java#L72-L74>`_ is the main driver, extracting the transaction list from each ``GetTransactionResponse``, and passing in to  ``processTransaction()`` for processing. This method, and the method ``processTransaction()`` implements the application logic.
```

File: README.rst [hunk 3]
```diff
@@ -64,22 +65,22 @@ The main function:
...
 
-`processTransaction() <src/main/java/examples/pingpong/grpc/PingPongProcessor.java#L98-L117>`_ extracts all creation events from the the transaction and passes them to ``processEvent()``. This produces a list of commands to be sent to the ledegr to further the workflow, and these are packages up in a ``Commands`` request and sent to the ledger.
+`processTransaction() <src/main/java/examples/pingpong/grpc/PingPongProcessor.java#L98-L117>`_ extracts all creation events from the the transaction and passes them to ``processEvent()``. This produces a list of commands to be sent to the ledger to further the workflow, and these are packages up in a ``Commands`` request and sent to the ledger.
```
Comment (atriantafyllos-da) [inline README.rst]:
```suggestion
`processTransaction() <src/main/java/examples/pingpong/grpc/PingPongProcessor.java#L96-L115>`_ extracts all creation events from the the transaction and passes them to ``processEvent()``. This produces a list of commands to be sent to the ledger to further the workflow, and these are packages up in a ``Commands`` request and sent to the ledger.
```

File: README.rst [hunk 4]
```diff
@@ -64,22 +65,22 @@ The main function:
...
 
 `processEvent() <src/main/java/examples/pingpong/grpc/PingPongProcessor.java#L129-L169>`_ takes a transaction event and turns it into a stream of commands to be sent back to the ledger. To do this, it examines the event for the correct package and template (it's a create of a ``Ping`` or ``Pong`` template) and then looks at the receiving part to decide if this processor should respond. If so, an exercise command for the correct choice is created and returned in a ``Stream``.
```
Comment (atriantafyllos-da) [inline README.rst]:
```suggestion
`processEvent() <src/main/java/examples/pingpong/grpc/PingPongProcessor.java#L127-L166>`_ takes a transaction event and turns it into a stream of commands to be sent back to the ledger. To do this, it examines the event for the correct package and template (it's a create of a ``Ping`` or ``Pong`` template) and then looks at the receiving part to decide if this processor should respond. If so, an exercise command for the correct choice is created and returned in a ``Stream``.
```

File: README.rst [hunk 5]
```diff
@@ -97,13 +98,13 @@ The first line shows that:
...
+Example Project -- Ping Pong with Reactive Components
+-----------------------------------------------------
 
 The code for this example is in the package `examples.pingpong.reactive <src/main/java/examples/pingpong/reactive>`_.
 
 PingPongReactiveMain.java
-^^^^^^^^^^^^^^^^^^^^^^^^^
+=========================
 
 The entry point for the Java code is the main class `PingPongReactiveMain <src/main/java/examples/pingpong/reactive/PingPongReactiveMain.java#L37-L82>`_.
```
Comment (atriantafyllos-da) [inline README.rst]:
```suggestion
The entry point for the Java code is the main class `PingPongReactiveMain <src/main/java/examples/pingpong/reactive/PingPongReactiveMain.java#L34-L80>`_.
```

File: README.rst [hunk 6]
```diff
@@ -120,8 +121,8 @@ At high level, the code does the following steps:
 - inject some contracts for each party of both templates
 - wait until the application is done
 
-PingPongProcessor.runIndefinitely()
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+PingPongProcessor.runIndefinitely
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 The core of the application is the method `PingPongProcessor.runIndefinitely() <src/main/java/examples/pingpong/reactive/PingPongProcessor.java#L42-L49>`_.
```
Comment (atriantafyllos-da) [inline README.rst]:
```suggestion
The core of the application is the method `PingPongProcessor.runIndefinitely() <src/main/java/examples/pingpong/reactive/PingPongProcessor.java#L43-L49>`_.
```

File: README.rst [hunk 7]
```diff
@@ -151,3 +152,72 @@ ReactiveX was chosen as the underlying library for the Java Binding because
...
+
+Example Project -- Ping Pong with Generated Java Data Layer
+-----------------------------------------------------------
+
+The code for this example is in the package  `examples.pingpong.codegen <src/main/java/examples/pingpong/codegen>`_.
+
+PingPongCodegenMain.java
+========================
+
+The entry point for the Java code is the main class `PingPongCodegenMain <src/main/java/examples/pingpong/codegen/PingPongCodegenMain.java#L46-L99>`_. Look at this class to see:
```
Comment (atriantafyllos-da) [inline README.rst]:
```suggestion
The entry point for the Java code is the main class `PingPongCodegenMain <src/main/java/examples/pingpong/codegen/PingPongCodegenMain.java#L35-L77>`_. Look at this class to see:
```

File: README.rst [hunk 8]
```diff
@@ -151,3 +152,72 @@ ReactiveX was chosen as the underlying library for the Java Binding because
...
+- creates an instance of a ``ManagedChannel`` connecting to an existing ledger
+- fetches the ledgerID and packageId from the ledger
+- creates ``Identifiers`` for the Ping and Pong templates
+- creates and starts instances of `PingPongProcessor <src/main/java/examples/pingpong/codegen/PingPongProcessor.java>`_ that contain the logic of the automation
+- injects the initial contracts to start the process
+
+PingPongProcessor.java
+======================
+
+The core of the application is the method `PingPongProcessor.runIndefinitely() <src/main/java/examples/pingpong/grpc/PingPongProcessor.java#L61-L91>`_.
```
Comment (atriantafyllos-da) [inline README.rst]:
```suggestion
The core of the application is the method `PingPongProcessor.runIndefinitely() <src/main/java/examples/pingpong/codegen/PingPongProcessor.java#L53-L87>`_.
```

File: README.rst [hunk 9]
```diff
@@ -151,3 +152,72 @@ ReactiveX was chosen as the underlying library for the Java Binding because
...
+
+This method retrieves a gRPC streaming endpoint using the ``GetTransactionsRequest`` request, and then creates a `RxJava <The Underlying Library: RxJava_>`_ ``StreamObserver``, providing implementations of the ``onNext``, ``onError`` and ``onComplete`` observer methods. ``RxJava`` arranges that these methods receive stream events asynchronously.
+
+The method `onNext <src/main/java/examples/pingpong/grpc/PingPongProcessor.java#L74-L76>`_ is the main driver, extracting the transaction list from each ``GetTransactionResponse``, and passing in to  ``processTransaction()`` for processing. This method, and the method ``processTransaction()`` implements the application logic.
```
Comment (atriantafyllos-da) [inline README.rst]:
```suggestion
The method `onNext <src/main/java/examples/pingpong/codegen/PingPongProcessor.java#L70-L72>`_ is the main driver, extracting the transaction list from each ``GetTransactionResponse``, and passing in to  ``processTransaction()`` for processing. This method, and the method ``processTransaction()`` implements the application logic.
```

File: README.rst [hunk 10]
```diff
@@ -151,3 +152,72 @@ ReactiveX was chosen as the underlying library for the Java Binding because
...
+
+`processTransaction() <src/main/java/examples/pingpong/grpc/PingPongProcessor.java#L98-L117>`_ extracts all creation events from the the transaction and passes them to ``processEvent()``. This produces a list of commands to be sent to the ledger to further the workflow, and these are packages up in a ``Commands`` request and sent to the ledger.
```
Comment (atriantafyllos-da) [inline README.rst]:
```suggestion
`processTransaction() <src/main/java/examples/pingpong/codegen/PingPongProcessor.java#L94-L110>`_ extracts all creation events from the the transaction and passes them to ``processEvent()``. This produces a list of commands to be sent to the ledger to further the workflow, and these are packages up in a ``Commands`` request and sent to the ledger.
```

File: README.rst [hunk 11]
```diff
@@ -151,3 +152,72 @@ ReactiveX was chosen as the underlying library for the Java Binding because
...
+The second line is analogous to the first one.
+
+The Generated Data Layer
+========================
+
+The ``codegen`` variant of the client application is similar to its ``grpc`` counterpart. Both are written in
+a traditional imperative style. What sets them apart is the usage of the generated data layer in the former.
+This layer simplifies construction of the ledger api calls and the analysis of the return values.
+
+- ``PingPongCodegenMain.createInitialContracts`` creates an strongly typed instance of a Ping contract and then embeds it in a equally strongly typed ``CommandsSubmission``. Then, it uses the built in ``toProto`` methods to convert the request into a wire-ready ``protobuf`` structure.
```
Comment (atriantafyllos-da) [inline README.rst]:
```suggestion
- ``PingPongCodegenMain.createInitialContracts`` creates a strongly typed instance of a Ping contract and then embeds it in an equally strongly typed ``CommandsSubmission``. Then, it uses the built in ``toProto`` methods to convert the request into a wire-ready ``protobuf`` structure.
```

File: README.rst [hunk 12]
```diff
@@ -151,3 +152,72 @@ ReactiveX was chosen as the underlying library for the Java Binding because
...
+
+The method `onNext <src/main/java/examples/pingpong/grpc/PingPongProcessor.java#L74-L76>`_ is the main driver, extracting the transaction list from each ``GetTransactionResponse``, and passing in to  ``processTransaction()`` for processing. This method, and the method ``processTransaction()`` implements the application logic.
+
+`processTransaction() <src/main/java/examples/pingpong/grpc/PingPongProcessor.java#L98-L117>`_ extracts all creation events from the the transaction and passes them to ``processEvent()``. This produces a list of commands to be sent to the ledger to further the workflow, and these are packages up in a ``Commands`` request and sent to the ledger.
+
+`processEvent() <src/main/java/examples/pingpong/grpc/PingPongProcessor.java#L129-L169>`_ takes a transaction event and turns it into a stream of commands to be sent back to the ledger. To do this, it examines the event for the correct package and template (it's a create of a ``Ping`` or ``Pong`` template) and then looks at the receiving part to decide if this processor should respond. If so, an exercise command for the correct choice is created and returned in a ``Stream``.
```
Comment (atriantafyllos-da) [inline README.rst]:
```suggestion
`processEvent() <src/main/java/examples/pingpong/codegen/PingPongProcessor.java#L122-L154>`_ takes a transaction event and turns it into a stream of commands to be sent back to the ledger. To do this, it examines the event for the correct package and template (it's a create of a ``Ping`` or ``Pong`` template) and then looks at the receiving part to decide if this processor should respond. If so, an exercise command for the correct choice is created and returned in a ``Stream``.
```

File: src/main/java/examples/pingpong/codegen/PingPongCodegenMain.java [hunk 13]
```diff
@@ -0,0 +1,128 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
```
Comment (atriantafyllos-da) [inline src/main/java/examples/pingpong/codegen/PingPongCodegenMain.java:2]: 2.0.0?

**Review by atriantafyllos-da — DISMISSED** [review]

File: src/main/java/examples/pingpong/codegen/PingPongCodegenMain.java (hunk 13 above)
Comment (mziolekda) [reply src/main/java/examples/pingpong/codegen/PingPongCodegenMain.java:2]: we use 2.0 throughout

**Review by atriantafyllos-da — APPROVED** [review]

**Pull request merged**